src/uct/tcp/tcp_cm.c

```diff
@@ -211,96 +208,149 @@ ucs_status_t uct_tcp_cm_send_event(uct_tcp_ep_t *ep,
+                                const struct sockaddr_in *dest_address,
+                                uct_tcp_cm_conn_sn_t conn_sn,
+                                uint8_t with_ctx_cap)
+{
+    ucs_conn_match_queue_type_t queue_type;
+    ucs_conn_match_elem_t *elem;
+    uct_tcp_ep_t *ep;
+    int remove_from_ctx;
+
+    ucs_assert(conn_sn < UCT_TCP_CM_CONN_SN_MAX);
+    ucs_assert((with_ctx_cap == UCT_TCP_EP_FLAG_CTX_TYPE_TX) ||
+               (with_ctx_cap == UCT_TCP_EP_FLAG_CTX_TYPE_RX));
+
+    /* Looking for EP with TX capability is a part of a connection request
+     * handling, the connection request could be received from a ghost
+     * connection that should be closed. So, need to search for the EP
+     * in any queue and don't remove EP in case of it was found */
+    if (with_ctx_cap == UCT_TCP_EP_FLAG_CTX_TYPE_TX) {
+        queue_type      = UCS_CONN_MATCH_QUEUE_ANY;
+        remove_from_ctx = 0;
+    } else {
+        queue_type      = UCS_CONN_MATCH_QUEUE_UNEXP;
+        remove_from_ctx = 1;
+    }
 
-    ucs_list_del(&ep->list);
-    ucs_debug("tcp_iface %p: tcp_ep %p removed from %p list",
-              iface, ep, ep_list);
+    elem = ucs_conn_match_get_elem(&iface->conn_match_ctx, dest_address,
+                                   conn_sn, queue_type, remove_from_ctx);
+    if (elem == NULL) {
+        return NULL;
+    }
 
-    uct_tcp_iface_add_ep(ep);
+    ep = ucs_container_of(elem, uct_tcp_ep_t, elem);
+    ucs_assert(ep->flags & (with_ctx_cap | UCT_TCP_EP_FLAG_ON_MATCH_CTX));
 
-    if (ucs_list_is_empty(ep_list)) {
-        kh_del(uct_tcp_cm_eps, &iface->ep_cm_map, iter);
-        ucs_debug("tcp_iface %p: %p list removed from map",
-                  iface, ep_list);
-        ucs_free(ep_list);
+    if (remove_from_ctx) {
+        ucs_assert((ep->flags & UCT_TCP_EP_CTX_CAPS) ==
+                   UCT_TCP_EP_FLAG_CTX_TYPE_RX);
+        ep->flags &= ~UCT_TCP_EP_FLAG_ON_MATCH_CTX;
+        uct_tcp_iface_add_ep(ep);
     }
+
+    return ep;
 }
 
-uct_tcp_ep_t *uct_tcp_cm_search_ep(uct_tcp_iface_t *iface,
-                                   const struct sockaddr_in *peer_addr,
-                                   uint8_t with_ctx_type)
+void uct_tcp_cm_insert_ep(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep)
 {
-    uct_tcp_ep_t *ep;
-    ucs_list_link_t *ep_list;
-    khiter_t iter;
+    uint8_t ctx_caps = ep->flags & UCT_TCP_EP_CTX_CAPS;
 
-    ucs_assert(with_ctx_type & UCT_TCP_EP_CTX_CAPS);
+    ucs_assert(ep->conn_sn < UCT_TCP_CM_CONN_SN_MAX);
+    ucs_assert((ctx_caps & UCT_TCP_EP_FLAG_CTX_TYPE_TX) ||
+               (ctx_caps == UCT_TCP_EP_FLAG_CTX_TYPE_RX));
+    ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_ON_MATCH_CTX));
 
-    iter = kh_get(uct_tcp_cm_eps, &iface->ep_cm_map, *peer_addr);
-    if (iter != kh_end(&iface->ep_cm_map)) {
-        ep_list = kh_value(&iface->ep_cm_map, iter);
-        ucs_assertv(!ucs_list_is_empty(ep_list), "iface=%p", iface);
+    uct_tcp_iface_remove_ep(ep);
```

Comment:
why was it added to the iface? if it's a new ep i'd expect it to be added directly to conn_match

Reply:
can't we add it to conn match immediately after creating?

Reply:
insert should only insert and not remove (can assert that the ep is not on the list when inserting)